Add Initial Mirage Node Type Support with Dedicated Configuration and Refined Init Checks

Masterix0 opened the pull request:
## Problem

The SKALE ecosystem requires support for a new node type, "Mirage," which operates differently from the standard SKALE node.
`node-cli` needs to be adapted to manage this new node type, including building a specific binary, providing dedicated CLI commands, handling Mirage-specific static parameters, and ensuring core functionalities correctly differentiate behavior, particularly around initialization and resource management.

## Solution

This pull request introduces foundational and refined support for the Mirage node type within `node-cli`.
It establishes a distinct build target for Mirage, introduces Mirage-specific CLI commands and core logic, and significantly refactors shared components to be node-type aware, especially in configuration handling and initialization.

**Key Changes Overview:**

1.  **Mirage Build Target & CLI Structure:**
    * Introduced build script and CI/CD workflow modifications to generate and publish a separate `mirage` binary.
    * Added new CLI command groups (`mirage boot`, `mirage node`) for the Mirage build type.
    * Implemented core logic for essential Mirage operations (boot init, migration, restore), with other commands reusing existing functionalities.

2.  **NodeType-Aware Architecture & Configuration Handling:**
    * Implemented a `NodeType` enum (`REGULAR`, `SYNC`, `MIRAGE`) to drive conditional logic throughout the codebase (node management, operations, Docker utilities, Nginx).
    * Introduced support for a dedicated `mirage_static_params.yaml` for Mirage-specific static configurations, allowing Mirage nodes to use standard `ENV_TYPE`s (e.g., `mainnet`, `devnet`) while sourcing distinct parameters. Core configuration loading (`get_static_params`) is now `NodeType`-aware.

3.  **Refined Initialization and Resource Management:**
    * Standardized the node initialization check (`is_node_inited`) to rely on the presence of `nginx.conf`, a common artifact for all initialized node types.
    * Ensured that Mirage-specific workflows correctly bypass SKALE's standard resource allocation calculations and shared BTRFS volume setup logic, aligning with Mirage's different resource model.

4.  **Testing & Documentation:**
    * Added and updated dedicated test suites for Mirage CLI commands, core logic, static parameter loading, `ENV_TYPE` validation, and initialization checks.
    * Updated the `README.md` to include instructions for the `mirage` binary, its commands, and its use of standard `ENV_TYPE`s with the new static parameter approach.


## Benefits

* **Extensibility:** Adds robust support for the Mirage node type, paving the way for future node types.
* **Clearer Configuration:** Mirage has its own clean static parameter file (`mirage_static_params.yaml`), simplifying management and avoiding bloat in the common `static_params.yaml`.
* **Maintainability:** Encapsulates Mirage-specific logic and uses `NodeType` for clearer control flow. Shared components like initialization checks and host requirement validation are now more universally applicable or correctly specialized.
* **Correctness:** Ensures Mirage nodes adhere to their specific operational model by bypassing inapplicable SKALE resource management schemes.
* **Alignment:** Integrates with corresponding `skale-admin` changes and the dedicated `docker-compose-mirage.yml`.
* **Test Coverage:** Includes dedicated and adapted tests for new Mirage functionality and configuration handling.
* **Usability:** Provides clear documentation for the `mirage` binary and its updated configuration paradigm.

## Testing Done

* Added new integration tests for Mirage CLI commands (mocking core layer).
* Added new unit/integration tests for Mirage core functions (mocking operations layer).
* Updated environment variable validation tests for Mirage types.
* Adapted existing core/check/utility tests to handle the `NodeType` parameter and verify behavior across types.
* CI workflows updated to build and test the Mirage binary type.

## Reviewer Instructions

1.  **High-Level Architecture:**
    * Focus on the overall strategy for supporting multiple node types via `NodeType`.
    * Assess the introduction of `mirage_static_params.yaml` and the `NodeType`-aware changes to `get_static_params`.
2.  **Mirage Workflow & Configuration:**
    * Conceptually review how Mirage nodes (boot and regular) source their configurations and initialize.
    * Verify that Mirage operations correctly bypass resource allocation and shared volume logic as intended.
    * Check the updated `ENV_TYPE` validation for Mirage.
3.  **Initialization Check:**
    * Review the updated `is_node_inited` logic and its impact on relevant decorators and commands.
4.  **Testing:**
    * Assess the updated test fixtures in `tests/conftest.py` for both static parameter files.
    * Ensure test adaptations in `core_checks_test.py`, `configs_env_validate_test.py`, and Mirage-specific tests correctly reflect the new configuration and initialization logic.
5.  **Documentation:**
    * Verify `README.md` updates, especially regarding `ENV_TYPE` usage for Mirage and any mention of static parameters.

Masterix0 committed:
Add initial Mirage node CLI structure and boot commands

Implements the foundational structure for the Mirage node type within the node-cli, including the build system integration and the core `mirage boot` commands.

Changes include:
- Added a new 'mirage' build type to `scripts/build.sh` and related scripts to generate the `skale-*-mirage` binary.
- Created new CLI files `node_cli/cli/mirage_boot.py` and `node_cli/cli/mirage_node.py`.
- Added the `mirage boot` command group with subcommands:
    - `init`: Implements logic to run the `mirage-boot` service from `docker-compose-mirage.yml`, ensuring `SKALE_NETWORK_TYPE=mirage` is set.
    - `register`: Reuses existing core registration logic.
    - `signature`: Reuses existing core signature logic.
    - `migrate`: Implements logic to switch from mirage boot to mirage main.
- Integrated the base `mirage` command group and existing `logs`, `wallet`, `ssl` groups into `main.py` for the `mirage` build type.
- Updated `README.md` with installation instructions for the `mirage` binary.

Masterix0 committed:
Implement mirage node commands (signature, backup, restore)

Completes the implementation of the `mirage node` command namespace by adding functionality for signature retrieval, backup, and restore operations.

Changes include:
- Implemented `mirage node signature` command, reusing the core `get_node_signature` function.
- Implemented `mirage node backup` command, reusing the core `backup` function.
- Added `restore_mirage` function to `node_cli/operations/base.py` to handle Mirage-specific restoration logic.
- Implemented `mirage node restore` command, utilizing the new `restore_mirage` function.
- Updated `node_cli/operations/base.py` functions (`turn_off`, `update_sync`) to use `NodeTypes` enum for better clarity (refactor).
- Fixed datetime formatting and typo in `node_cli/core/node.py`.
- Removed placeholder file `node_cli/operations/mirage_boot.py` as logic is integrated elsewhere.

Placeholders remain for `mirage node init` and `mirage node register`.

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -300,7 +311,7 @@ def backup(path):
 
 def get_backup_filename():
-    time = datetime.datetime.utcnow().strftime('%Y-%m-%d-%H-%M-%S')
+    time = datetime.datetime.now(datetime.timezone.utc).strftime('%Y-%m-%d-%H-%M-%S')
```
double-check that it returns data in the same format as before

dmytrotkk commented on node_cli/cli/mirage_node.py:
```diff
@@ -0,0 +1,79 @@
+)
+@streamed_cmd
+def restore_node(backup_path, env_file, no_snapshot, config_only):
+    restore_mirage(backup_path, env_file, no_snapshot, config_only)
```
add a placeholder for `mirage update` cmd

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -224,17 +234,18 @@ def compose_node_env(
+        env_params = get_validated_env_config(INIT_ENV_FILEPATH, node_type=node_type)
 
-    mnt_dir = SCHAINS_MNT_DIR_SYNC if sync_node else SCHAINS_MNT_DIR_REGULAR
+    mnt_dir = SCHAINS_MNT_DIR_SYNC if node_type == NodeTypes.SYNC else SCHAINS_MNT_DIR_REGULAR
```
we will have to update this one to use `SCHAINS_MNT_DIR_SYNC` for mirage nodes too (probably makes sense to rename it)

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -404,11 +415,25 @@ def turn_on(maintenance_off, sync_schains, env_file):
+
+
+def is_base_containers_alive(node_type: NodeTypes = NodeTypes.REGULAR) -> bool:
+    if node_type == NodeTypes.MIRAGE:
```
we need to think if it makes sense to rename all container name prefixes to `mirage_`, since most of the containers will actually stay the same (transaction-manager, api, redis, etc)

dmytrotkk commented on node_cli/operations/base.py:
```diff
@@ -149,7 +148,47 @@ def update(env_filepath: str, env: Dict) -> bool:
-def init(env_filepath: str, env: dict) -> bool:
+def migrate_mirage_boot(env_filepath: str, env: Dict) -> bool:
+    compose_rm(env, node_type=NodeTypes.MIRAGE)
+    remove_dynamic_containers()
```
it's critical to DO NOT stop dynamic containers during the migration

dmytrotkk commented on node_cli/operations/base.py:
```diff
@@ -233,12 +302,11 @@ def init_sync(
 
 def update_sync(env_filepath: str, env: Dict) -> bool:
-    compose_rm(env, sync_node=True)
+    compose_rm(env, node_type=NodeTypes.MIRAGE)
```
why `node_type=NodeTypes.MIRAGE` in `update_sync`?

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -96,11 +101,20 @@ class NodeStatuses(Enum):
-        return True
-    if sync_node and not is_sync_admin_running():
-        return True
+class NodeTypes(Enum):
```
suggest renaming this to `NodeType`:

> PEP 8 recommends CapWords (PascalCase) for class names and using singular nouns for enums (unless the enum is explicitly a collection of plural entities, which is rare and discouraged).

also, docstring can be dropped

dmytrotkk commented on node_cli/core/mirage_node.py:
```diff
@@ -0,0 +1,54 @@
+
+    if not no_snapshot:
+        logger.info('Adding BACKUP_RUN to env ...')
+        env['BACKUP_RUN'] = 'True'
```
no need for BACKUP_RUN in mirage node restart - mirage nodes always start from snapshot

dmytrotkk commented on node_cli/core/nginx.py:
```diff
@@ -34,10 +34,12 @@
 
 def generate_nginx_config() -> None:
     ssl_on = check_ssl_certs()
+    regular_node = TYPE != 'mirage'
```
use `NodeType` for comparison; mb create a separate func `is_regular_node`

dmytrotkk commented on node_cli/utils/docker_utils.py:
```diff
@@ -54,6 +55,18 @@
     'filebeat',
 )
+BASE_SYNC_COMPOSE_SERVICES = ('skale-sync-admin', 'nginx')
+PROTO_MIRAGE_COMPOSE_SERVICES = (
```
mb let's change to `CORE_` for better clarity

dmytrotkk commented on node_cli/utils/docker_utils.py:
```diff
@@ -322,16 +377,24 @@ def is_container_running(name: str, dclient: Optional[DockerClient] = None) -> b
+def is_admin_running(
+    node_type: NodeTypes = NodeTypes.REGULAR, client: Optional[DockerClient] = None
+) -> bool:
+    if node_type == NodeTypes.MIRAGE:
```
should it also handle `mirage_boot_admin` case?

Masterix0 replied:
I don't think so, it's only used to check is it's safe to update in update_op (which we don't have for boot) and we don't check if it's safe to update when we migrate.
I just added mirage node support so we can use it in the future when implementing the rest of the CLI ops.

dmytrotkk commented on node_cli/utils/meta.py:
```diff
@@ -19,12 +20,14 @@ def __new__(
         os_version=DEFAULT_OS_VERSION,
     ):
+        # If docker_lvmpy_stream is None, use the default value
+        actual_docker_lvmpy_stream = docker_lvmpy_stream or DEFAULT_DOCKER_LVMPY_STREAM
```
what is the purpose of this line?
isn’t it already covered by `docker_lvmpy_stream: Optional[str] = DEFAULT_DOCKER_LVMPY_STREAM`?

Masterix0 replied:
It handles passing 'None' explicitly. But maybe it's an unnecessary edge case that doesn't happen. WIll remove.

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -96,11 +101,20 @@ class NodeStatuses(Enum):
+class NodeTypes(Enum):
+    """This class contains possible node types."""
+
+    REGULAR = 0
```
does it make sense to generate info.py in following format now?

```python
from node_cli.core.node import NodeTypes

BUILD_DATETIME = '2025-04-07 17:28:11'
COMMIT = '36844c629974ec606850277fc5cb6c1f91cc8aee'
BRANCH = 'develop'
OS = 'Darwin-arm64'
VERSION = '2.6.0'
TYPE = NodeTypes.REGULAR

```

Masterix0 committed:
Addressed review comments for PR #851 Draft

Masterix0 committed:
Updated README to include Mirage commands and to have better formatting. Removed some unused options from mirage commands.

Masterix0 committed:
Eliminated the step to make the pre-commit hook executable, as we'll need to investigate pre-commit in future.

Masterix0 committed:
Fix circular import for NodeType

Build broke because info.py need NodeType from core.node,
but core.node need env.py, and env.py need NodeType from core.node.
Make circle. Bad.

Move NodeType to utils/node_types.py.
Update imports in core.node, configs.env, generate_info.sh.

No more circle. Build work now.

Masterix0 committed:
Fixes incorrect NodeType import in `docker_utils.py`

Masterix0 committed:
Replaced sync_node flag with node_type in build_env_params.

Masterix0 committed:
Updated helper-scripts

Masterix0 committed:
Updated README Development section with more complete install instructions and fixed pre-commit hook

Masterix0 committed:
Updated static params handling by Checkers to ignore requirements that have value equal to 'disabled'

Masterix0 committed:
Refactored `is_update_safe` tests with improved coverage.

Masterix0 committed:
Fixed `is_update_safe` tests mock path

Masterix0 committed:
Changed `is_base_containers_alive` to check both `mirage_` and `skale_` prefixes for mirage node

Masterix0 committed:
Reworked env.py tests; Introduced explicit node type parameters to related functions.

Masterix0 committed:
Refactored functions to explicitly require the NodeType parameter, removing default values.

Masterix0 committed:
Started fixing env validation tests

Masterix0 committed:
Removed unused import in env tests

Masterix0 committed:
Fixed broken tests; Adapted `validate_env_params` to work with params required by new node types

Masterix0 committed:
Added mirage build and verification steps to github actions test workflow.

Masterix0 committed:
Initial implementation of new core_node tests

Masterix0 committed:
Fixed updated `core_node_test.py`

Masterix0 committed:
Updated docker-compose checks test; Added mirage disabled package requirements test

Masterix0 committed:
Added simple `info` command CLI test

Masterix0 committed:
Added nginx template processing tests.

Masterix0 committed:
Added mirage core tests

Masterix0 committed:
Added mirage CLI tests

Masterix0 committed:
Fixed unused import

dmytrotkk commented on .github/workflows/publish.yml:
```diff
@@ -178,3 +178,62 @@ jobs:
         asset_name: ${{ matrix.asset_name }}.sha512
         asset_content_type: text/plain
+
+  build_and_publish_mirage:
```
let's check if it's possible to optimize normal/sync/mirage publish pipelines to avoid code duplication

maybe something similar to https://github.com/skalenetwork/skale.py/blob/c29d3a3812cb62b2804de8342e805fec69266008/.github/workflows/test.yml#L8

dmytrotkk commented on node_cli/cli/node.py:
```diff
@@ -166,7 +179,7 @@ def remove_node_from_maintenance():
 @streamed_cmd
 def _turn_off(maintenance_on, unsafe_ok):
-    turn_off(maintenance_on, unsafe_ok)
+    turn_off(node_type=TYPE, maintenance_on=maintenance_on, unsafe_ok=unsafe_ok)
```
why `node_type=NodeType.REGULAR,` is used in some places and `node_type=TYPE` in others?

badrogger commented on node_cli/cli/mirage_node.py:
```diff
@@ -0,0 +1,93 @@
+@streamed_cmd
+def update_node(env_file, pull_config_for_schain, unsafe_ok):
+    click.echo("Placeholder: Command 'mirage node update' is not yet implemented.")
+    pass
```
should be removed

Masterix0 replied:
What should be removed? The update command placeholder? D1 explicitly told me to add it, if I'm not mistaken.

badrogger replied:
The pass statement should be removed.

badrogger commented on node_cli/configs/env.py:
```diff
@@ -24,34 +24,47 @@
 
-ALLOWED_ENV_TYPES = ['mainnet', 'testnet', 'qanet', 'devnet']
+ALLOWED_SKALE_ENV_TYPES = ['mainnet', 'testnet', 'qanet', 'devnet']
+ALLOWED_MIRAGE_ENV_TYPES = ['mainnet-mirage', 'devnet-mirage']
```
It seems that separation here is redundant. Both networks will have mainnet/testnet/qanet/devnet.

Masterix0 replied:
It is not redundant. Since they share the same config file, the environment types and names are distinct for Skale/Mirage. Not sure if I misunderstood the comment.

badrogger commented on node_cli/core/mirage_boot.py:
```diff
@@ -0,0 +1,78 @@
+    if not is_base_containers_alive(node_type=NodeType.MIRAGE, is_mirage_boot=True):
+        error_exit('Containers are not running', exit_code=CLIExitCodes.OPERATION_EXECUTION_ERROR)
+    logger.info('Generating mirage resource allocation file ...')
+    update_resource_allocation(env['ENV_TYPE'])
```
If resource allocation is still required for mirage the generate_allocation cli should support NodeType.Mirage.

Masterix0 replied:
I don't remember whether this is required and is supposed to be here or if it is an error. Would have to confirm with D1.

dmytrotkk replied:
let's drop resource allocation from the mirage node

badrogger commented on node_cli/core/nginx.py:
```diff
@@ -47,6 +50,10 @@ def check_ssl_certs():
 
 
+def is_regular_node_nginx() -> bool:
+    return TYPE != NodeType.MIRAGE
```
Sync node type for MIRAGE will be added in the future. It's better to check that TYPE in [NodeType.SKALE, NodeType.SYNC]

badrogger commented on node_cli/core/node.py:
```diff
@@ -379,37 +401,57 @@ def set_maintenance_mode_off():
 
-def is_base_containers_alive(sync_node: bool = False):
+def get_expected_container_names(node_type: NodeType, is_mirage_boot: bool) -> list[str]:
+    if node_type == NodeType.MIRAGE and is_mirage_boot:
```
It might be cleaner to return only once at the end of the function. 
if node_type == TypeA:
     services = ServicesA
elif:
    services = ServicesB      
     
return list(services.values())

badrogger commented on node_cli/core/node.py:
```diff
@@ -379,37 +401,57 @@ def set_maintenance_mode_off():
+    running_container_names = [container.name for container in dclient.containers.list()]
+
+    for base_container in base_container_names:
+        if base_container in running_container_names:
```
if base_container not in running_container_names:
     return False
     
 running_container_names can be set() instead of a list.

badrogger commented on tests/core/nginx_test.py:
```diff
@@ -0,0 +1,144 @@
+    os.makedirs(os.path.dirname(NGINX_TEMPLATE_FILEPATH), exist_ok=True)
+    with open(NGINX_TEMPLATE_FILEPATH, 'w') as f:
+        f.write(TEST_NGINX_TEMPLATE)
+    yield
```
yield should be surrounded by try/except.
also it is better to check os.path.isfile than catching FileNotFoundError.

badrogger commented on node_cli/utils/docker_utils.py:
```diff
@@ -322,16 +387,20 @@ def is_container_running(name: str, dclient: Optional[DockerClient] = None) -> b
 
-def is_sync_admin_running(dclient: Optional[DockerClient] = None) -> bool:
-    return is_container_running(name='skale_sync_admin', dclient=dclient)
+def is_admin_running(node_type: NodeType, client: Optional[DockerClient] = None) -> bool:
```
Better to leave only one return statement.

badrogger commented on node_cli/utils/docker_utils.py:
```diff
@@ -245,47 +270,87 @@ def compose_rm(env={}, sync_node: bool = False):
+        return COMPOSE_PATH
+
+
+def get_compose_services(node_type: NodeType) -> list[str]:
```
Better to leave only one return statement.

Masterix0 committed:
Addressed some comments in PR #851

Masterix0 committed:
Continued addressing comments from PR # 851.
Removed pass statements

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -24,34 +24,47 @@
 
-ALLOWED_ENV_TYPES = ['mainnet', 'testnet', 'qanet', 'devnet']
+ALLOWED_SKALE_ENV_TYPES = ['mainnet', 'testnet', 'qanet', 'devnet']
+ALLOWED_MIRAGE_ENV_TYPES = ['mainnet-mirage', 'devnet-mirage']
```
remove mirage env types

Masterix0 committed:
Continued addressing comments in PR # 851

Masterix0 committed:
Refactored publish pipeline to avoid code duplication

Masterix0 committed:
Added `update` command to mirage CLI

Masterix0 committed:
Fixed unnecessary arguments

Masterix0 committed:
Updated codebase to use mirage_static_params.yaml for Mirage nodes.

- Updated static param loading to be NodeType-aware.
- Normalized ENV_TYPEs for Mirage.
- Bypassed resource allocation for Mirage init/restore.
- Updated is_node_inited to use nginx.conf.

Masterix0 committed:
Removed unexistent import

Masterix0 committed:
Fixed `configs_env_validate_test.py` tests

Masterix0 committed:
Continued work in fixing tests after mirage static params changes

Masterix0 committed:
Fixed all tests that were broken due to mirage static params

dmytrotkk commented on text.yml:
```diff
@@ -76,3 +76,11 @@ lvmpy:
+
+mirage:
+  init:
+    help: Initialize sync SKALE node
```
why sync?

dmytrotkk commented on text.yml:
```diff
@@ -76,3 +76,11 @@ lvmpy:
+mirage:
+  init:
+    help: Initialize sync SKALE node
+    indexer: Run sync node in indexer mode (disable block rotation)
```
no such options in mirage init

dmytrotkk commented on tests/core/core_node_test.py:
```diff
@@ -9,54 +9,211 @@
+    mock_get_validated.return_value = valid_env_params.copy()
+    if node_type == NodeType.SYNC:
+        mock_get_validated.return_value['ENV_TYPE'] = 'devnet'
+    elif node_type == NodeType.MIRAGE:
```
this elif can be removed

dmytrotkk commented on tests/core/core_mirage_test.py:
```diff
@@ -0,0 +1,130 @@
+    resource_alloc,
+    meta_file_v3,
+):
+    mock_env = {'ENV_TYPE': 'devnet-mirage'}
```
env type doesn't exist but test is still passing?

Masterix0 replied:
Yes, it's mocked. I'll fix it to a correct one, though

dmytrotkk commented on tests/core/core_mirage_test.py:
```diff
@@ -0,0 +1,130 @@
+    valid_env_file,
+    ensure_meta_removed,
+):
+    mock_env = {'ENV_TYPE': 'devnet-mirage'}
```
non-existing env type

dmytrotkk commented on tests/conftest.py:
```diff
@@ -109,6 +111,197 @@
+
 
 @pytest.fixture
 def net_params_file():
```
rename to env_params_file and mirage_env_params_file

dmytrotkk commented on tests/conftest.py:
```diff
@@ -109,6 +111,197 @@
     docker-engine: 1.1.3
 """
 
+MIRAGE_TEST_ENV_PARAMS = """
```
any reason to keep it in the code?

Masterix0 committed:
Addressed comments for PR #851

Masterix0 committed:
Fixed unused imports
